File: backend/schemas/game.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime
from backend.models.game import GameStatus


class ThrowCreate(BaseModel):
    scores: List[int] = Field(..., min_length=1, max_length=3)
    multipliers: Optional[List[Optional[int]]] = Field(None, min_length=1, max_length=3)


class ThrowResponse(BaseModel):
    id: UUID
    game_id: UUID
    player_id: UUID
    turn_number: int
    scores: List[int]
    multipliers: Optional[List[Optional[int]]]
    total: int
    remaining: Optional[int]
    is_bust: bool
    created_at: datetime

    class Config:
        from_attributes = True


class GameCreate(BaseModel):
    match_id: UUID
    set_number: int = Field(..., ge=1)
    leg_number: int = Field(..., ge=1)


class GameUpdate(BaseModel):
    status: Optional[GameStatus] = None
    current_player_id: Optional[UUID] = None
    winner_id: Optional[UUID] = None
    game_data: Optional[Dict[str, Any]] = None


class GameResponse(BaseModel):
    id: UUID
    match_id: UUID
    set_number: int
    leg_number: int
    status: GameStatus
    current_player_id: Optional[UUID]
    winner_id: Optional[UUID]
    game_data: Dict[str, Any]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ScoreSubmission(BaseModel):
    game_id: UUID
    player_id: UUID
    throw: ThrowCreate

File: shared/types/game.ts
export enum GameStatus {
  PENDING = "pending",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
}

export interface Game {
  id: string;
  match_id: string;
  set_number: number;
  leg_number: number;
  status: GameStatus;
  current_player_id?: string;
  winner_id?: string;
  game_data: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Throw {
  id: string;
  game_id: string;
  player_id: string;
  turn_number: number;
  scores: number[];
  multipliers?: (number | null)[];
  total: number;
  remaining?: number;
  is_bust: boolean;
  created_at: string;
}

export interface ThrowCreate {
  scores: number[];
  multipliers?: (number | null)[];
}

export interface ScoreSubmission {
  game_id: string;
  player_id: string;
  throw: ThrowCreate;
}
